src/knn.py

```diff
@@ -0,0 +1,60 @@
+"""Implement a K-Nearest Neighbors aglorithm."""
+import pandas as pd
+from math import sqrt
+
+
+class KNearestNeighbors(object):
+    """Define a K-Nearest Neighbors object."""
+
+    def __init__(self, data, k=5):
+        """Initialize a k nearest neighbors object."""
+        if type(k) is not int or k <= 0:
+            raise ValueError("Please initalize with positive integer.")
+        else:
+            self.k = k
+        if type(data) is not pd.DataFrame:
+            try:
+                self.data = pd.DataFrame(data)
+            except pd.PandasError:
+                pass
+        else:
+            self.data = data
+
+    def predict(self, test_data, tk=None):
+        """Given data, categorize the data by its k nearest neighbors."""
+        if tk is None:
+            tk = self.k
+        if type(test_data) is not pd.DataFrame:
+            try:
+                test_data = pd.Series(test_data)
+            except pd.PandasError:
+                raise ValueError("BAD DATA YA TURKEY")
+        distances = []
+        for row in self.data.iterrows():
+            distances.append((row[1][-1], self._distance(row[1], test_data)))
+        distances.sort(key=lambda x: x[1])
+        # import pdb; pdb.set_trace()
+        my_class = self._classify(distances[:tk])
+        if my_class:
+            return my_class
+        else:
+            self.predict(test_data, tk - 1)
```

Comment:
Written for the case in which the classification is a "tie" between two classes.  In that case, the classify function returns None and therefore predict is run once again with a decreased k value.  This is per the description of the algorithm from the class notes.

https://codefellows.github.io/sea-python-401d5/lectures/k_nearest_neighbors.html?highlight=nearest